src/libsystemd/sd-device/device-trigger.c

```diff
@@ -0,0 +1,483 @@
+
+        sd_device_trigger_callback_t pre_trigger_cb;
+        void *pre_trigger_cb_userdata;
+        sd_device_trigger_callback_t settle_cb;
+        void *settle_cb_userdata;
+
+        bool executing:1;
+};
+
+static int device_trigger_ensure_uuid_set(sd_device_trigger *trigger) {
+        sd_id128_t id;
+        int r;
+
+        if (!isempty(trigger->uuid))
+                return 0;
+
+        r = sd_id128_randomize(&id);
+        if (r < 0)
+                return log_debug_errno(r, "sd-device-trigger: Failed to generate random UUID: %m");
+
+        id128_to_uuid_string(id, trigger->uuid);
+        return 0;
+}
+
+static inline void device_trigger_discard_uuid(sd_device_trigger *trigger) {
+        trigger->uuid[0] = '\0';
+}
+
+_public_ int sd_device_trigger_new(sd_device_trigger **ret) {
+        _cleanup_(sd_device_trigger_unrefp) sd_device_trigger *trigger = NULL;
+
+        assert(ret);
+
+        trigger = new(sd_device_trigger, 1);
+        if (!trigger)
+                return -ENOMEM;
+
+        *trigger = (sd_device_trigger) {
+                .n_ref = 1,
+                .action = device_action_to_string(DEVICE_ACTION_CHANGE),
+                .source = DEFAULT_SOURCE_VALUE,
+        };
+
+        device_trigger_discard_uuid(trigger);
+
+        *ret = TAKE_PTR(trigger);
+
+        return 0;
+}
+
+static sd_device_trigger *device_trigger_free(sd_device_trigger *trigger) {
+        size_t i;
+
+        for (i = 0; i < trigger->n_devices; i++)
+                sd_device_unref(trigger->devices[i]);
+
+        free(trigger->devices);
+        hashmap_free_free_free(trigger->properties);
+
+        return mfree(trigger);
+}
+
+DEFINE_PUBLIC_TRIVIAL_REF_UNREF_FUNC(sd_device_trigger, sd_device_trigger, device_trigger_free);
+
+_public_ int sd_device_trigger_set_action(sd_device_trigger *trigger, const char *action) {
+        DeviceAction device_action;
+
+        assert_return(trigger, -EINVAL);
+
+        if (trigger->executing)
+                return -EPERM;
+
+        device_action = device_action_from_string(action);
+        if (device_action < 0)
+                return -EINVAL;
+
+        trigger->action = action;
+
+        return 0;
+}
+
+_public_ const char *sd_device_trigger_get_action(sd_device_trigger *trigger) {
+        assert_return_errno(trigger, NULL, EINVAL);
+
+        return trigger->action;
+}
+
+_public_ const char *sd_device_trigger_get_uuid(sd_device_trigger *trigger) {
+        int r;
+
+        assert_return(trigger, NULL);
+
+        r = device_trigger_ensure_uuid_set(trigger);
+        if (r < 0)
+                return NULL;
+
+        return trigger->uuid;
+}
+
+_public_ int sd_device_trigger_set_source(sd_device_trigger *trigger, const char *source) {
+        assert_return(trigger, -EINVAL);
+
+        if (trigger->executing)
+                return -EPERM;
+
+        if (!in_charset(source, ALPHANUMERICAL))
+                return -EINVAL;
+
+        trigger->source = source;
```

Comment:
hmm, a ptr to a user-supplied string? that is weird... please make this robust and copy the thing